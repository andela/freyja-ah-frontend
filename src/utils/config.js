const image = 'https://cdn.umnbootcamp.com/wp-content/uploads/sites/93/2018/03/placeholder-person.png';
const baseUrl = 'https://freyja-ah-backend.herokuapp.com/api';

export {
  image,
  baseUrl,
};
