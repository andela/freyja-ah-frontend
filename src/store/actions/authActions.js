import axios from 'axios';
import jwtDecode from 'jwt-decode';
import { GET_ERRORS, SET_CURRENT_USER, LOGIN_ERROR, INIT_AUTH_REQUEST, END_AUTH_REQUEST, VERIFY_USER } from './types';
import setAuthToken from '../../utils/setAuthToken';
import { baseUrl } from '../../utils/config';


/**
 * @method setCurrentUser
 * @description sets the current user of the app
 * @param {object} decoded user details
  * @returns {object}
 */
export const setCurrentUser = decoded => ({
  type: SET_CURRENT_USER,
  payload: decoded,
});

/**
 * @method setError
 * @description sets app error
 * @param {object} error
  * @returns {object}
 */
export const setError = error => ({
  type: GET_ERRORS,
  errors: error,
});

export const verifyUser = state => ({
  type: VERIFY_USER,
  isVerified: state,
});

/**
 * @method registerUser
 * @description registers a new user
 * @param {object} newUser
  * @returns {object} user object
 */
export const registerUser = newUser => async (dispatch) => {
  try {
    const user = await axios.post(`${baseUrl}/users`, newUser);

    if (user) {
      const { token } = user.data;
      localStorage.setItem('token', token);
      const decoded = jwtDecode(token);
      dispatch(setCurrentUser(decoded));
    }
  } catch (err) {
    const { error } = err.response.data;
    dispatch(setError(error));
  }
};

/**
 * @method loginUser
 * @description registers a new user
 * @param {object} payload
  * @returns {object} user object
 */
export const loginUser = payload => async (dispatch) => {
  dispatch({ type: INIT_AUTH_REQUEST });

  try {
    const request = await axios.post(`${baseUrl}/users/login`, payload);
    dispatch({
      type: END_AUTH_REQUEST,
    });

    const { token } = request.data;
    localStorage.setItem('token', token);

    setAuthToken(token);
    const decoded = jwtDecode(token);
    dispatch(setCurrentUser(decoded));
  } catch (error) {
    dispatch({
      type: END_AUTH_REQUEST,
    });

    const { data } = error.response;
    const errorObj = Array.isArray(data.error) ? { error: data.error[0].msg } : data;
    dispatch({
      type: LOGIN_ERROR,
      payload: errorObj,
    });
  }
};


/**
 * @method verifyAuthUser
 * @description registers a new user
 * @param {string} token
  * @returns {string} confirmation message
 */
export const verifyAuthUser = token => async (dispatch) => {
  try {
    await axios.get(`${baseUrl}/user/verify/${token}`);
    await dispatch(verifyUser("you've been verified, Redirecting to your profile page..."));
  } catch (error) {
    dispatch(verifyUser("Couldn't verify you, Please chcek the link sent to your email again."));
  }
};
