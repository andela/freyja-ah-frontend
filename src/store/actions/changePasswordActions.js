import axios from 'axios';
import { baseUrl } from '../../utils/config';
import { PASSWORD_CHANGE_ERROR, PASSWORD_CHANGE_SUCCESS, LOADING } from './types';

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method passwordChangeSuccess
 * @description handles password change success
 * @param {string} data success message
  * @returns {object}
 */
export const passwordChangeSuccess = data => ({
  type: PASSWORD_CHANGE_SUCCESS,
  payload: data,
});

/**
 * @method passwordChangeError
 * @description handles password change error
 * @param {string} data error message
  * @returns {object}
 */
export const passwordChangeError = data => ({
  type: PASSWORD_CHANGE_ERROR,
  payload: data,
});

/**
 * @method ChangePassword
 * @description change user's password
 * @param {string} password
 * @param {string} token
  * @returns {object}
 */
export const ChangePassword = (password, token) => async (dispatch) => {
  try {
    dispatch(loading());

    const resetPassword = await axios.post(
      `${baseUrl}/users/change-password?token=${token}`,
      { newPassword: password },
    );

    dispatch(passwordChangeSuccess(resetPassword.data.message));
  } catch (error) {
    dispatch(passwordChangeError('please send reset password request again'));
  }
};
