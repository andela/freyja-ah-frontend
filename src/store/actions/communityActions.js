import { get } from 'axios';
import { GET_COMMUNITY_MESSAGES, GET_COMMUNITY_MESSAGES_ERROR, LOADING } from './types';
import { baseUrl } from '../../utils/config';

/**
 * @method communityMessageSuccess
 * @param {string} data success message
  * @returns {object}
 */
export const communityMessageSuccess = data => ({
  type: GET_COMMUNITY_MESSAGES,
  data,
});

/**
 * @method communityMessageSuccess
 * @param {string} data error message
  * @returns {object}
 */
export const communityMessageError = data => ({
  type: GET_COMMUNITY_MESSAGES_ERROR,
  data,
});

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method communityMessageSuccess
 * @description get all community messages
  * @returns {object}
 */
export const getCommunityMessages = () => async (dispatch) => {
  dispatch(loading());
  try {
    const response = await get(`${baseUrl}/community/messages`, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });
    dispatch(communityMessageSuccess(response.data.data));
  } catch (error) {
    dispatch(communityMessageError('internal error'));
  }
};
