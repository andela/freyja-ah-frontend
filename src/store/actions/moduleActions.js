import { get } from 'axios';
import {
  GET_ALL_MODULES, GET_SINGLE_MODULE, MODULES_ERROR, INIT_MODULE_REQUEST,
} from './types';
import { baseUrl } from '../../utils/config';

/**
 * @method getModules
 * @description get all course modules
  * @returns {object}
 */
export const getModules = () => async (dispatch) => {
  try {
    const modulesRequest = await get(`${baseUrl}/modules`, {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });

    const { data } = modulesRequest.data;
    return dispatch({
      type: GET_ALL_MODULES,
      payload: data,
    });
  } catch (error) {
    return dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};

/**
 * @method setModules
 * @description set current module
 * @param {object} payload module object
  * @returns {object}
 */
export const setModule = payload => ({
  type: GET_SINGLE_MODULE,
  payload,
});

export const getModule = moduleId => async (dispatch) => {
  try {
    dispatch({ type: INIT_MODULE_REQUEST });
    const request = await get(`${baseUrl}/modules/${moduleId}`);
    const { data } = request.data;
    dispatch(setModule(data));
  } catch (error) {
    dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};
