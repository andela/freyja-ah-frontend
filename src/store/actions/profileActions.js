import axios from 'axios';
import * as actionCreators from './types';
import { baseUrl } from '../../utils/config';

const token = localStorage.getItem('token');

/**
 * @method getProfileSuccess
 * @param {object} profile profile object
  * @returns {object}
 */
export const getProfileSuccess = profile => ({
  type: actionCreators.GET_PROFILE_SUCCESS,
  data: profile,
});

/**
 * @method getProfileFailed
 * @param {string} error profile object
  * @returns {object}
 */
export const getProfileFailed = error => ({
  type: actionCreators.GET_PROFILE_FAILED,
  error,
});

/**
 * @method  getProfile
 * @param {string} userId user id
  * @returns {object}
 */
export const getProfile = userId => async (dispatch) => {
  const config = {
    headers: {
      Authorization: token,
    },
  };
  try {
    const profile = await axios.get(`${baseUrl}/profiles/${userId}`, config);
    const { data } = profile;
    const { user } = data;
    let dob = null;
    if (user.profile.dateOfBirth) {
      const dobArray = (user.profile.dateOfBirth).split('T');
      [dob] = dobArray;
    }

    const userDetails = {
      name: `${user.firstName} ${user.lastName}`,
      email: user.email,
      bio: user.profile.bio,
      image: user.profile.image,
      username: user.userName,
      phoneNumber: user.profile.phoneNumber,
      socialMedia: [
        { name: 'facebook', url: user.profile.facebook },
        { name: 'twitter', url: user.profile.twitter },
        { name: 'linkedin', url: user.profile.linkedIn },
      ],
      DOB: dob,
      industry: user.profile.industry,
      yrsOfExperience: user.profile.yrsOfExperience,
      isEnrolled: user.profile.isEnrolled,
    };

    dispatch(getProfileSuccess(userDetails));
  } catch (error) {
    dispatch(getProfileFailed(error));
  }
};

/**
 * @method  uploadStart
 * @description initiates image upload
  * @returns {object}
 */
export const uploadStart = () => ({
  type: actionCreators.UPLOAD_START,
});

/**
 * @method  uploadSuccess
 * @description handles image upload success
 * @param {string} imageUrl upload image url
  * @returns {object}
 */
export const uploadSuccess = imageUrl => ({
  type: actionCreators.UPLOAD_SUCCESS,
  imageUrl,
});

/**
 * @method  uploadFailed
 * @description handles image upload failure
 * @param {object} error error object
  * @returns {object}
 */
export const uploadFailed = error => ({
  type: actionCreators.UPLOAD_FAILED,
  error,
});

/**
 * @method  uploadImage
 * @description handles image upload
 * @param {file} image image
  * @returns {object}
 */
export const uploadImage = image => async (dispatch) => {
  dispatch(uploadStart());
  const config = {
    headers: {
      'Content-type': 'multipart/form-data',
      Authorization: token,
    },
  };
  try {
    const data = new FormData();
    data.append('image', image);

    const upload = await axios.post(`${baseUrl}/image`, data, config);
    const imageUrl = upload.data.data;
    dispatch(uploadSuccess(imageUrl));
  } catch (error) {
    dispatch(uploadFailed(error.response.data));
  }
};

/**
 * @method  updateProfileFailed
 * @description handles profile update failure
 * @param {object} error update profile error
  * @returns {object}
 */
export const updateProfileFailed = error => ({
  type: actionCreators.UPDATE_PROFILE_FAILED,
  error,
});

/**
 * @method  updateProfile
 * @description handles profile update
 * @param {object} data profile data
  * @returns {object}
 */
export const updateProfile = data => async (dispatch) => {
  const config = {
    headers: {
      Authorization: token,
    },
  };
  try {
    const profile = await axios.put(`${baseUrl}/profiles`, data, config);
    await dispatch(getProfile(profile.data.profile.userId));
  } catch (error) {
    dispatch(updateProfileFailed(error.response.data.error));
  }
};
