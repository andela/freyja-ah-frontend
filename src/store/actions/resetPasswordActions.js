import axios from 'axios';
import { baseUrl } from '../../utils/config';
import { PASSWORD_RESET_SUCCESS, PASSWORD_RESET_ERROR, LOADING } from './types';

/**
 * @method loading
 * @description initiates loading
  * @returns {object}
 */
export const loading = () => ({
  type: LOADING,
});

/**
 * @method passwordResetSuccess
 * @description handles password reset success
 * @param {string} data success message
  * @returns {object}
 */
export const passwordResetSuccess = data => ({
  type: PASSWORD_RESET_SUCCESS,
  payload: data,
});

/**
 * @method passwordResetError
 * @description handles password reset error
 * @param {string} data error message
  * @returns {object}
 */
export const passwordResetError = data => ({
  type: PASSWORD_RESET_ERROR,
  payload: data,
});

/**
 * @method ResetPassword
 * @description reset user's password
 * @param {string} email
  * @returns {object}
 */
export const ResetPassword = email => async (dispatch) => {
  try {
    dispatch(loading());

    const resetPassword = await axios.post(`${baseUrl}/users/reset`, { email });

    dispatch(passwordResetSuccess(resetPassword.data.message));
  } catch (error) {
    dispatch(passwordResetError(error.response.data.message));
  }
};
