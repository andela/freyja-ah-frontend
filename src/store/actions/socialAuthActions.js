import jwtDecode from 'jwt-decode';
import { setCurrentUser, setError } from './authActions';
import { baseUrl } from '../../utils/config';

export const socialAuth = token => async (dispatch) => {
  try {
    localStorage.setItem('token', token);
    const decoded = jwtDecode(token);
    dispatch(setCurrentUser(decoded));
  } catch (error) {
    const { err } = error.response.data;
    dispatch(setError(err));
  }
};


export const socialAuthPath = (authType) => {
  window.location.href = `${baseUrl}/auth/${authType}`;
};

export const getToken = (tokenString) => {
  const startIndex = tokenString.indexOf('=') + 1;
  const token = tokenString.slice(startIndex);
  return token;
};
