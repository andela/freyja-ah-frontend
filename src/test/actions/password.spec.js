import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import {
  ResetPassword,
  loading,
  passwordResetSuccess,
  passwordResetError,
} from '../../store/actions/resetPasswordActions';
import {
  ChangePassword,
  passwordChangeSuccess,
  passwordChangeError,
} from '../../store/actions/changePasswordActions';
import { testBaseUrl } from '../config/testConfig';
import { verifyAuthUser, verifyUser } from '../../store/actions/authActions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('creates PASSWORD_RESET_SUCCESS', (done) => {
    nock(`${testBaseUrl}`)
      .post('/api/users/reset')
      .reply(200, { message: 'Please check your email' });

    store.dispatch(ResetPassword('mattimobolaji@gmail.com')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
    done();
  });

  it('creates PASSWORD_RESET_ERROR', (done) => {
    nock(`${testBaseUrl}`).post('/api/users/reset').reply(401, { error: 'invalid email' });

    store.dispatch(ResetPassword('mattimobolaji@gmail.commm')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
    done();
  });

  it('creates PASSWORD_CHANGE_SUCCESS when', () => {
    nock(`${testBaseUrl}`)
      .post('api/users/change-password?token=amdacsdacmmaddd')
      .reply(200, { message: 'password updated successfully' });

    return store.dispatch(ChangePassword('december', 'cadjsdasscdadadsds')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('verifies user', (done) => {
    nock(`${testBaseUrl}`).get('api/user/verify/jdkkdkkd').reply(200, {});

    store.dispatch(verifyAuthUser('dnsanda')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
    done();
  });

  it('update verified state', () => {
    store.dispatch(verifyUser(true));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates CHANGE_PASSWORD_SUCCESS when change password action is successful', () => {
    store.dispatch(passwordChangeSuccess('password successfully update'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates CHANGE_PASSWORD_ERROR when change password action is not successful', () => {
    store.dispatch(passwordChangeError('sorry, check the link again'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates PASSWORD_RESET_SUCCESS when reset password action is successful', () => {
    store.dispatch(passwordResetSuccess('Please check your email'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('creates PASSWORD_RESET_ERROR when reset password action is not successful', () => {
    store.dispatch(passwordResetError('invalid Email'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('updates loading to true in state', () => {
    store.dispatch(loading());
    expect(store.getActions()).toMatchSnapshot();
  });
});
