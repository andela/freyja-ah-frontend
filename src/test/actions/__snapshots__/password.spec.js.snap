// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates CHANGE_PASSWORD_ERROR when change password action is not successful 1`] = `
Array [
  Object {
    "payload": "sorry, check the link again",
    "type": "PASSWORD_CHANGE_ERROR",
  },
]
`;

exports[`async actions creates CHANGE_PASSWORD_SUCCESS when change password action is successful 1`] = `
Array [
  Object {
    "payload": "password successfully update",
    "type": "PASSWORD_CHANGE_SUCCESS",
  },
]
`;

exports[`async actions creates PASSWORD_CHANGE_SUCCESS when 1`] = `
Array [
  Object {
    "type": "LOADING",
  },
  Object {
    "payload": "please send reset password request again",
    "type": "PASSWORD_CHANGE_ERROR",
  },
]
`;

exports[`async actions creates PASSWORD_RESET_ERROR when reset password action is not successful 1`] = `
Array [
  Object {
    "payload": "invalid Email",
    "type": "PASSWORD_RESET_ERROR",
  },
]
`;

exports[`async actions creates PASSWORD_RESET_SUCCESS when reset password action is successful 1`] = `
Array [
  Object {
    "payload": "Please check your email",
    "type": "PASSWORD_RESET_SUCCESS",
  },
]
`;

exports[`async actions update verified state 1`] = `
Array [
  Object {
    "isVerified": true,
    "type": "VERIFY_USER",
  },
]
`;

exports[`async actions updates loading to true in state 1`] = `
Array [
  Object {
    "type": "LOADING",
  },
]
`;
